data-tx/src/main/java/io/micronaut/transaction/interceptor/TransactionalInterceptor.java

```diff
@@ -91,55 +91,74 @@ public int getOrder() {
+            switch (interceptedMethod.resultType()) {
+                case PUBLISHER:
+                    return interceptedMethod.handleResult(
+                            transactionInvocation.reactiveTransactionOperations.withTransaction(definition, (status) -> {
+                                context.setAttribute(ReactiveTransactionStatus.STATUS, status);
+                                context.setAttribute(ReactiveTransactionStatus.ATTRIBUTE, definition);
+                                return Publishers.convertPublisher(context.proceed(), Publisher.class);
+                            })
+                    );
+                case COMPLETION_STAGE:
+                    if (transactionInvocation.reactiveTransactionOperations != null) {
+                        return interceptedMethod.handleResult(interceptedMethod.interceptResult());
+                    } else {
+                        throw new ConfigurationException("Async return type doesn't support transactional execution.");
+                    }
+                case SYNCHRONOUS:
+                    final SynchronousTransactionManager<?> transactionManager = transactionInvocation.transactionManager;
+                    final TransactionInfo transactionInfo = createTransactionIfNecessary(
+                            transactionManager,
+                            definition,
+                            context.getExecutableMethod()
+                    );
+                    Object retVal;
+                    try {
+                        retVal = context.proceed();
+                    } catch (Throwable ex) {
+                        completeTransactionAfterThrowing(transactionInfo, ex);
+                        throw ex;
+                    } finally {
+                        cleanupTransactionInfo(transactionInfo);
+                    }
+                    commitTransactionAfterReturning(transactionInfo);
+                    return retVal;
+                default:
+                    return interceptedMethod.unsupported();
 
-                return new TransactionInvocation<>(transactionManager, null, transactionAttribute);
-            }
-        });
-
-        final TransactionAttribute definition = transactionInvocation.definition;
-        if (transactionInvocation.reactiveTransactionOperations != null) {
-            return Publishers.convertPublisher(transactionInvocation.reactiveTransactionOperations.withTransaction(definition, (status) -> {
-                context.setAttribute(ReactiveTransactionStatus.STATUS, status);
-                context.setAttribute(ReactiveTransactionStatus.ATTRIBUTE, definition);
-                return Publishers.convertPublisher(context.proceed(), Publisher.class);
-            }), returnType.getType());
-        } else {
-            final SynchronousTransactionManager<?> transactionManager = transactionInvocation.transactionManager;
```

Comment:
Why did you explicitly remove support for proceeding with transactions for async methods?

Now, this results in a `ConfigurationException("Async return type doesn't support transactional execution.");`.

This got our project into a bit of upgrade pains as we used to (I suppose, improperly?) rely on suspend functions in bog-standard CrudRepositories.

Reply:
I didn't remove it, previously coroutines were executed synchronously so we were able to call `SynchronousTransactionManager` but now calling it with async code wouldn't work, there needs to be a different solution

Reply:
Ah! It was fake async! That makes a lot of sense given `SynchronousTransactionManager` being here.

Thank you for taking the time to answer my question.